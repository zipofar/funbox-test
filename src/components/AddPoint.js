import React from 'react';

export default class AddPoint extends React.Component
{
    state = {
        inputValue: '',
        searchedPoints: [],
        currentPoint: [],
        currentIndexSearch: null,
        hoverIndex: null,
    };

    onChangeValue = (e) => {
        const { value } = e.target;
        this.setState({ inputValue: value, currentIndexSearch: null });
        this.props.searchPoints(value, (points) => {this.setState({ searchedPoints: points })});
    };

    resetState = () => {
        this.setState({ inputValue: '', searchedPoints: [],  currentPoint: [] });
    };

    savePoint = () => {
        this.props.addPoint(this.state.currentPoint);
        this.resetState();
    };

    onSubmitPoint = (e) => {
        e.preventDefault();
        this.savePoint();
    };

    showDropDownSearch = () => {
        const { searchedPoints } = this.state;
        if (searchedPoints.length === 0) {
            return null;
        }
        return(
            <div className='col-sm-12'>
                <div id="searchedPoints" className='form-row'>
                    <div className="list-group">
                        { searchedPoints.map((item, i) => {
                            return(
                                <button
                                    key={i}
                                    type='button'
                                    className={`list-group-item ${ this.state.hoverIndex === i ? 'active' : '' }`}
                                    onClick={this.onClickDropDownPoints(item.displayName, i)}
                                    onMouseEnter={this.toggleHoverFindedItems(i)}
                                    onMouseLeave={this.toggleHoverFindedItems(i)}
                                >
                                    {item.displayName}
                                </button>
                            );
                        }) }
                    </div>
                </div>
            </div>
        );
    };

    toggleHoverFindedItems = index => (e) => {
        if (index === this.state.hoverIndex) {
            this.setState({ hoverIndex: null });
        } else {
            this.setState({
                hoverIndex: index,
                inputValue: this.state.searchedPoints[index].displayName,
                currentPoint: this.state.searchedPoints[index],
                currentIndexSearch: index,
            });
        }
    };

    onClickDropDownPoints = (value, index) => (e) => {
        e.preventDefault();
        this.setState({
            inputValue: value,
            currentPoint: this.state.searchedPoints[index],
            searchedPoints: [],
        }, this.savePoint);
    };

    onBlurPointInput = () => {
        setTimeout(() => this.resetState(), 400);
    };

    keyUpInputPoint = (e) => {
        if (e.key === 'Escape') {
            this.resetState();
        }
        if (e.key !== 'ArrowDown' && e.key !== 'ArrowUp') {
            return;
        }
        if (this.state.searchedPoints.length === 0) {
            return;
        }

        let currentIndex = this.state.currentIndexSearch;

        if (e.key === 'ArrowDown') {
            if (currentIndex === null) {
                currentIndex = 0;
            } else if (this.state.searchedPoints.length - 1 > currentIndex) {
                currentIndex += 1;
            }
        }

        if (e.key === 'ArrowUp') {
            if (currentIndex === null || currentIndex === 0) {
                return;
            } else if (currentIndex > 0) {
                currentIndex -= 1;
            }
        }

        this.setState({
            inputValue: this.state.searchedPoints[currentIndex].displayName,
            currentPoint: this.state.searchedPoints[currentIndex],
            hoverIndex: currentIndex,
            currentIndexSearch: currentIndex,
        });
    };

    render() {
        return(
            <div className='row'>
                <form onSubmit={this.onSubmitPoint} style={{width: '100%'}} autoComplete='off'>
                    <div className='form-row'>
                        <div className='col-sm-12'>
                            <input
                                type="text"
                                className="form-control"
                                id="inputPoint"
                                placeholder="New point route"
                                onChange={this.onChangeValue}
                                onBlur={this.onBlurPointInput}
                                onKeyUp={this.keyUpInputPoint}
                                value={this.state.inputValue}
                            />
                        </div>
                        {this.showDropDownSearch()}
                    </div>
                </form>
            </div>
        );
    }
}