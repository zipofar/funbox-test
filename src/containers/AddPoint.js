import { connect } from 'react-redux';
import AddPoint from '../components/AddPoint';
import * as actions from '../actions';

const mapStateToProps = state => {
    return {

    };
};

const container = connect(mapStateToProps, actions)(AddPoint);

export default container;