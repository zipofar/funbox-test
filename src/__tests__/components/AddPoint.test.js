import React from 'react';
import { shallow, configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { shallowToJson } from 'enzyme-to-json';
import AddPoint from '../../components/AddPoint';

configure({ adapter: new Adapter() });

it('render AddPoint without crashing', () => {
    shallow(<AddPoint />);
});

it('write some symbol and call func "searchPoints"', () => {
    const searchPoints = jest.fn();
    const comp = shallow(<AddPoint searchPoints={searchPoints}/>);
    const input = comp.find('#inputPoint');
    input.prop('onChange')({ target: { value: 'a' } });
    expect(searchPoints).toBeCalled();
});

describe('DropDownSearch element', () => {
    let comp;

    const searchedPoints = [
        { displayName: 'Moscow'},
        { displayName: 'Penza'},
        { displayName: 'Sochi'},
    ];

    beforeEach(() => {
        comp = shallow(<AddPoint />);
    });

    it('Should render list finded addresses', () => {
        comp.setState({ searchedPoints, });
        expect(shallowToJson(comp)).toMatchSnapshot();
    });

    it('Should toggle state.hoverIndex when mouse hover on list finded addresses', () => {
        comp.setState({ searchedPoints, });

        const firstPoint = comp.find('button.list-group-item').at(0);
        firstPoint.simulate('mouseenter');
        expect(comp.state().hoverIndex).toEqual(0);

        firstPoint.simulate('mouseleave');
        expect(comp.state().hoverIndex).toEqual(null);
    });

    it('Should add and remove class "active" when mouse hover on element of list finded addresses', () => {

        //onMouseEnter
        comp.setState({ searchedPoints, hoverIndex: 0 });
        const firstAddressInactive = comp.find('button.list-group-item').at(0);
        expect(shallowToJson(firstAddressInactive)).toMatchSnapshot();

        //onMouseLeave
        comp.setState({ searchedPoints, hoverIndex: null });
        const firstAddressActive = comp.find('button.list-group-item').at(0);
        expect(shallowToJson(firstAddressActive)).toMatchSnapshot();

    });

    it('Click on searched point element must change state', () => {
        comp = mount(<AddPoint addPoint={ () => {} } />);
        comp.setState({ searchedPoints, });

        const firstPoint = comp.find('button.list-group-item').at(0);
        firstPoint.simulate('click', {
            preventDefault: () => {},
        });
        expect(comp.state().inputValue).toEqual('');
        expect(comp.state().currentPoint).toEqual([]);
        expect(comp.state().searchedPoints).toEqual([]);
    });

    it('Submit form must change state', () => {

        comp = mount(<AddPoint addPoint={ () => {} } />);
        comp.setState({ searchedPoints, });

        // Click on first point
        const firstPoint = comp.find('button.list-group-item').at(0);
        firstPoint.simulate('click', {
            preventDefault: () => {},
        });

        //Click on submit
        const form = comp.find('form');
        form.simulate('submit', {
            preventDefault: () => {},
        });

        expect(comp.state().inputValue).toEqual('');
        expect(comp.state().currentPoint).toEqual([]);
        expect(comp.state().searchedPoints).toEqual([]);
        expect(shallowToJson(comp)).toMatchSnapshot();

    });

    it('KeyDown and KeyUp on InputPoint must toggle class on selected point', () => {
        comp = mount(<AddPoint />);

        comp.setState({ searchedPoints, });

        const inputPoint = comp.find('#inputPoint');

        //Activated first address
        inputPoint.simulate('keyUp', { key: 'ArrowDown' });
        const searchedPoints1 = comp.find('#searchedPoints');
        expect(shallowToJson(searchedPoints1)).toMatchSnapshot();

        //Activated second address
        inputPoint.simulate('keyUp', { key: 'ArrowDown' });
        const searchedPoints2 = comp.find('#searchedPoints');
        expect(shallowToJson(searchedPoints2)).toMatchSnapshot();

        //Again activated first address
        inputPoint.simulate('keyUp', { key: 'ArrowUp' });
        const searchedPoints3 = comp.find('#searchedPoints');
        expect(shallowToJson(searchedPoints3)).toMatchSnapshot();

        //Activated third address like a onMouseEnter
        comp.setState({ hoverIndex: 2, currentIndexSearch: 2 });
        const searchedPoints4 = comp.find('#searchedPoints');
        expect(shallowToJson(searchedPoints4)).toMatchSnapshot();

        //Activated second address
        inputPoint.simulate('keyUp', { key: 'ArrowUp' });
        const searchedPoints5 = comp.find('#searchedPoints');
        expect(shallowToJson(searchedPoints5)).toMatchSnapshot();

        //Activated first address like a onMouseEnter
        comp.setState({ hoverIndex: 0, currentIndexSearch: 0 });
        const searchedPoints6 = comp.find('#searchedPoints');
        expect(shallowToJson(searchedPoints6)).toMatchSnapshot();

        //Activated second address
        inputPoint.simulate('keyUp', { key: 'ArrowDown' });
        const searchedPoints7 = comp.find('#searchedPoints');
        expect(shallowToJson(searchedPoints7)).toMatchSnapshot();
    });

});
