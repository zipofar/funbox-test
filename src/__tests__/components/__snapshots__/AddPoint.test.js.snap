// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropDownSearch element KeyDown and KeyUp on InputPoint must toggle class on selected point 1`] = `
<div
  className="form-row"
  id="searchedPoints"
>
  <div
    className="list-group"
  >
    <button
      className="list-group-item active"
      key="0"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      type="button"
    >
      Moscow
    </button>
    <button
      className="list-group-item "
      key="1"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      type="button"
    >
      Penza
    </button>
    <button
      className="list-group-item "
      key="2"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      type="button"
    >
      Sochi
    </button>
  </div>
</div>
`;

exports[`DropDownSearch element KeyDown and KeyUp on InputPoint must toggle class on selected point 2`] = `
<div
  className="form-row"
  id="searchedPoints"
>
  <div
    className="list-group"
  >
    <button
      className="list-group-item "
      key="0"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      type="button"
    >
      Moscow
    </button>
    <button
      className="list-group-item active"
      key="1"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      type="button"
    >
      Penza
    </button>
    <button
      className="list-group-item "
      key="2"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      type="button"
    >
      Sochi
    </button>
  </div>
</div>
`;

exports[`DropDownSearch element KeyDown and KeyUp on InputPoint must toggle class on selected point 3`] = `
<div
  className="form-row"
  id="searchedPoints"
>
  <div
    className="list-group"
  >
    <button
      className="list-group-item active"
      key="0"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      type="button"
    >
      Moscow
    </button>
    <button
      className="list-group-item "
      key="1"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      type="button"
    >
      Penza
    </button>
    <button
      className="list-group-item "
      key="2"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      type="button"
    >
      Sochi
    </button>
  </div>
</div>
`;

exports[`DropDownSearch element KeyDown and KeyUp on InputPoint must toggle class on selected point 4`] = `
<div
  className="form-row"
  id="searchedPoints"
>
  <div
    className="list-group"
  >
    <button
      className="list-group-item "
      key="0"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      type="button"
    >
      Moscow
    </button>
    <button
      className="list-group-item "
      key="1"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      type="button"
    >
      Penza
    </button>
    <button
      className="list-group-item active"
      key="2"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      type="button"
    >
      Sochi
    </button>
  </div>
</div>
`;

exports[`DropDownSearch element KeyDown and KeyUp on InputPoint must toggle class on selected point 5`] = `
<div
  className="form-row"
  id="searchedPoints"
>
  <div
    className="list-group"
  >
    <button
      className="list-group-item "
      key="0"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      type="button"
    >
      Moscow
    </button>
    <button
      className="list-group-item active"
      key="1"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      type="button"
    >
      Penza
    </button>
    <button
      className="list-group-item "
      key="2"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      type="button"
    >
      Sochi
    </button>
  </div>
</div>
`;

exports[`DropDownSearch element KeyDown and KeyUp on InputPoint must toggle class on selected point 6`] = `
<div
  className="form-row"
  id="searchedPoints"
>
  <div
    className="list-group"
  >
    <button
      className="list-group-item active"
      key="0"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      type="button"
    >
      Moscow
    </button>
    <button
      className="list-group-item "
      key="1"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      type="button"
    >
      Penza
    </button>
    <button
      className="list-group-item "
      key="2"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      type="button"
    >
      Sochi
    </button>
  </div>
</div>
`;

exports[`DropDownSearch element KeyDown and KeyUp on InputPoint must toggle class on selected point 7`] = `
<div
  className="form-row"
  id="searchedPoints"
>
  <div
    className="list-group"
  >
    <button
      className="list-group-item "
      key="0"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      type="button"
    >
      Moscow
    </button>
    <button
      className="list-group-item active"
      key="1"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      type="button"
    >
      Penza
    </button>
    <button
      className="list-group-item "
      key="2"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      type="button"
    >
      Sochi
    </button>
  </div>
</div>
`;

exports[`DropDownSearch element Should add and remove class "active" when mouse hover on element of list finded addresses 1`] = `
<button
  className="list-group-item active"
  key="0"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  type="button"
>
  Moscow
</button>
`;

exports[`DropDownSearch element Should add and remove class "active" when mouse hover on element of list finded addresses 2`] = `
<button
  className="list-group-item "
  key="0"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  type="button"
>
  Moscow
</button>
`;

exports[`DropDownSearch element Should render list finded addresses 1`] = `
<div
  className="row"
>
  <form
    autoComplete="off"
    onSubmit={[Function]}
    style={
      Object {
        "width": "100%",
      }
    }
  >
    <div
      className="form-row"
    >
      <div
        className="col-sm-12"
      >
        <input
          className="form-control"
          id="inputPoint"
          onBlur={[Function]}
          onChange={[Function]}
          onKeyUp={[Function]}
          placeholder="New point route"
          type="text"
          value=""
        />
      </div>
      <div
        className="col-sm-12"
      >
        <div
          className="form-row"
          id="searchedPoints"
        >
          <div
            className="list-group"
          >
            <button
              className="list-group-item "
              key="0"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
              type="button"
            >
              Moscow
            </button>
            <button
              className="list-group-item "
              key="1"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
              type="button"
            >
              Penza
            </button>
            <button
              className="list-group-item "
              key="2"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
              type="button"
            >
              Sochi
            </button>
          </div>
        </div>
      </div>
    </div>
  </form>
</div>
`;

exports[`DropDownSearch element Submit form must change state 1`] = `
<AddPoint
  addPoint={[Function]}
>
  <div
    className="row"
  >
    <form
      autoComplete="off"
      onSubmit={[Function]}
      style={
        Object {
          "width": "100%",
        }
      }
    >
      <div
        className="form-row"
      >
        <div
          className="col-sm-12"
        >
          <input
            className="form-control"
            id="inputPoint"
            onBlur={[Function]}
            onChange={[Function]}
            onKeyUp={[Function]}
            placeholder="New point route"
            type="text"
            value=""
          />
        </div>
      </div>
    </form>
  </div>
</AddPoint>
`;
